File: skills/say_skill/say_skill.cc
#include "say_skill.h"

#include <memory>

#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/synchronization/notification.h"
#include "skills/say_skill/say_skill.pb.h"
#include "intrinsic/util/status/status_macros.h"
#include "intrinsic/skills/cc/skill_utils.h"
#include "intrinsic/skills/proto/skill_service.pb.h"

namespace say_skill {

using ::com::example::SaySkillParams;

using ::intrinsic::skills::ExecuteRequest;
using ::intrinsic::skills::SkillInterface;
using ::intrinsic::skills::ExecuteContext;

// -----------------------------------------------------------------------------
// Skill signature.
// -----------------------------------------------------------------------------

std::unique_ptr<SkillInterface> SaySkill::CreateSkill() {
  return std::make_unique<SaySkill>();
}

// -----------------------------------------------------------------------------
// Skill execution.
// -----------------------------------------------------------------------------

absl::StatusOr<std::unique_ptr<google::protobuf::Message>> SaySkill::Execute(
    const ExecuteRequest& request, ExecuteContext& context) {

  // Get parameters.
  INTR_ASSIGN_OR_RETURN(
    auto params, request.params<SaySkillParams>());

  context.canceller().Ready();
  if (context.canceller().Wait(absl::Milliseconds(params.wait_ms()))) {
    LOG(INFO) << "say_skill_cc was canceled!";
    return absl::CancelledError("say_skill_cc was canceled.");
  }

  LOG(INFO) << params.text();

  return nullptr;
}
}  // namespace say_skill

File: skills/say_skill/say_skill.py
"""Contains the skill say_skill_py."""

from absl import logging

from intrinsic.skills.python import skill_interface
from intrinsic.util.decorators import overrides


class SaySkill(skill_interface.Skill):
    """Implementation of the say_skill_py skill."""

    def __init__(self) -> None:
        pass

    @overrides(skill_interface.Skill)
    def execute(self, request: skill_interface.ExecuteRequest,
                context: skill_interface.ExecuteContext
                ) -> None:

        context.canceller.ready()
        if context.canceller.wait(request.params.wait_ms / 1000.0):
            raise skill_interface.SkillCancelledError(
                f"say_skill_cc was cancelled."
            )

        logging.info(request.params.text)
